code/modules/effects/map_effects/airlock_marker.dm

```diff
@@ -0,0 +1,135 @@
+
+/// Specialization helper for the airlock marker, to be put above "interior" parts of the airlock,
+/// and on top of the actual airlock marker. By itself does nothing.
+/obj/effect/map_effect/marker_helper/airlock/interior
+	name = "airlock marker (interior/inside/pressurized)"
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "airlock_marker_interior"
+	layer = LIGHTING_LAYER
+
+/// add the airlock market to `airlock_markers`
+/obj/effect/map_effect/marker/airlock/Initialize(mapload, ...)
+	..()
+	if(master_tag && frequency)
+		if(!airlock_markers[master_tag])
+			airlock_markers[master_tag] = list()
+		airlock_markers[master_tag] += src
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/effect/map_effect/marker/airlock/LateInitialize()
+	if(master_tag && frequency)
+		airlock_marker_init_airlock(master_tag)
+		airlock_markers[master_tag] = null // init only once
+
+#define MASTER_TAG		"[master_tag]_controller"
+#define AIRPUMP_TAG		"[master_tag]_pump"
+#define SENSOR_TAG		"[master_tag]_sensor"
+#define EXTERIOR_DOOR_TAG	"[master_tag]_outer"
+#define INTERIOR_DOOR_TAG	"[master_tag]_inner"
+
+/// actually set the airlock up
+/proc/airlock_marker_init_airlock(var/master_tag)
+	// if airlock of this master_tag was already initialized, return
+	if(!airlock_markers || !airlock_markers[master_tag] || airlock_markers[master_tag]==null)
+		return
+
+	// iterate through every marker of this airlock and set up the component parts
+	for(var/obj/effect/map_effect/marker/airlock/marker in airlock_markers[master_tag])
+		var/is_interior = locate(/obj/effect/map_effect/marker_helper/airlock/interior) in marker.loc
+		var/is_exterior = locate(/obj/effect/map_effect/marker_helper/airlock/exterior) in marker.loc
+
+		var/frequency = marker.frequency
+		var/required_access = marker.required_access
+
+		if(frequency==null || master_tag==null)
+			return
+
+		for(var/thing in marker.loc)
+			var/obj/machinery/embedded_controller/radio/airlock/airlock_controller/airlock_controller = thing
+			if(istype(airlock_controller))
+				airlock_controller.set_frequency(frequency)
+				airlock_controller.id_tag = MASTER_TAG
+				airlock_controller.tag_airpump = AIRPUMP_TAG
+				airlock_controller.tag_chamber_sensor = SENSOR_TAG
+				airlock_controller.tag_exterior_door = EXTERIOR_DOOR_TAG
+				airlock_controller.tag_interior_door = INTERIOR_DOOR_TAG
+				airlock_controller.program = new /datum/computer/file/embedded_program/airlock(airlock_controller)
+				airlock_controller.req_access = required_access
+
+			var/obj/machinery/door/airlock/door = thing
+			if(istype(door))
+				door.set_frequency(frequency)
+				door.req_access = required_access
+				door.lock()
+				if(is_interior)
+					door.id_tag = INTERIOR_DOOR_TAG
+				if(is_exterior)
+					door.id_tag = EXTERIOR_DOOR_TAG
+
+			var/obj/machinery/airlock_sensor/sensor = thing
+			if(istype(sensor))
+				sensor.set_frequency(frequency)
+				sensor.id_tag = SENSOR_TAG
+				sensor.master_tag = MASTER_TAG
```

Comment:
this should `continue`, as types in this list of checks are mutually exclusive, there's no reason to continue the typechecks once one is found

Reply:
fixed